### ISSUE TITLE
Rules that should be refined further

### ISSUE THREAD

**Issue by magurotuna** [denoland/deno_lint#330]
Now that all the recommended rules have been implemented, it's good time to look back at the rules again.
I have looked through how the rules are implemented, and have noticed that some rules need more refinement. Current issues I've found are:

- unable to catch violations in nested blocks, just like #327 
- some cases that should be reported are ignored (e.g. normal funcions are considered, but _arrow_ functions are not)
- insufficient test cases (I think porting the ESLint's tests would be good as it would guarantee the compatibility with ESLint.)

### rules that need refinement

Note that this list has been made from my quick look, thus it may contain rules that actually don't need any fix.

- [x] default-param-last #329 
- [ ] explicit-function-return-type
- [ ] for-direction
- [ ] getter-return
- [ ] no-array-constructor
- [ ] no-async-promise-executor
- [ ] no-case-declarations
- [ ] no-class-assign
- [ ] no-compare-neg-zero
- [ ] no-cond-assign
- [ ] no-const-assign
- [ ] no-constant-condition
- [ ] no-dupe-args
- [ ] no-dupe-keys
- [ ] no-duplicate-case
- [ ] no-empty
- [ ] no-eval
- [ ] no-ex-assign
- [ ] no-func-assign
- [ ] no-global-assign
- [ ] no-inferrable-type
- [ ] no-invalid-regexp
- [ ] no-new-symbol
- [ ] no-non-null-asserted-optional-chain
- [ ] no-obj-calls
- [ ] no-prototype-builtins
- [ ] no-self-assign
- [ ] no-setter-return
- [ ] no-shadow-restricted-names
- [ ] no-sparse-arrays
- [ ] no-this-alias
- [ ] no-throw-literal
- [x] no-unsafe-finally #328 
- [ ] no-unsafe-negation
- [ ] no-var
- [ ] no-with
- [ ] prefer-as-const
- [ ] single-var-declarator
- [ ] use-isnan
- [ ] valid-typeof

**Comment by bartlomieju** [comment]
@magurotuna have you found bugs in all the listed rules? 

IIRC most of implemented rules have ESLint tests ported, but I'm more than happy to add more tests if they are lacking.

**Comment by magurotuna** [comment]
@bartlomieju 

> have you found bugs in all the listed rules?

not necessarily, but I feel like lots of the listed rules are unable to catch nested violations.
For example, with `no-with` rule, if we run a test like

```rust
    assert_lint_err_n::<NoWith>(
      "with (foo) { with(bar) { console.log(42); }}",
      vec![0, 13],
    );
```

it is expected to pass, but the rule doesn't detect `with(bar)`. This may go for the other rules.

I will fix them one by one, giving priority to "recommended" rules over the others. Hopefully it will not take much time, but I guess it will...

**Comment by bartlomieju** [comment]
@magurotuna I see, thanks for digging into this problem.

**Comment by bartlomieju** [comment]
@magurotuna is this issue still relevant?

**Comment by magurotuna** [comment]
@bartlomieju Not really sure, let me check if the issue remains.
